### PR TITLE
UX improvement on Google account card

### PR DESCRIPTION
**Pull request by ecgan**
### Changes proposed in this Pull Request:

Part of #1065 .

This PR contains UX improvement on Google account card in Setup MC Step 1, and allows users to connect to a different Google account after connecting to an existing account.

### Screenshots:

Disconnected card:

image

Request more access card:

image

Connected card with the ability to connect to a different Google account: 

image

🎥   Demo video with voice showing switching to another Google account: (https://d.pr/v/1NoXVS)

https://user-images.githubusercontent.com/417342/140094226-e411b0fb-03f8-479e-8e85-fbd7456ecff1.mov

### Detailed test instructions:

Pre-condition: you should have multiple Google accounts to test switching accounts.

1. Disconnect Google account in connection test page: `/wp-admin/admin.php?page=connection-test-admin-page`
2. Go through the Setup MC flow. You should see the disconnected card with Connect button.
3. Click on the Connect button. In the Google permission page, select only the "Manage your product listings and accounts for Google Shopping" permission and click Allow.
4. When you are back in Setup MC, you should see the request more access card with "Allow full access" button. 
5. Click on the button. In the Google permission page, click Allow.
6. When you are back in Setup MC, you should see the connected card.
7. Click on Connect to a different Google account. In the Google page, select another Google account. Grant permissions if needed.
8. When you are back in Setup MC, you should see the connected card with the different Google account.

### Changelog entry

> Update - UX improvement on Google account card in Setup MC.
> Add - Allow connecting to a different Google account in Setup MC.

### CODE REVIEW

File: js/src/components/google-account-card/connected-google-account-card.js [hunk 1]
```diff
@@ -1,14 +1,58 @@
...
+		} catch ( error ) {
+			createNotice(
+				'error',
+				__(
+					'Unable to connect your Google account. Please try again later.',
+					'google-listings-and-ads'
+				)
+			);
+		}
+	};
```
Comment (ecgan) [inline js/src/components/google-account-card/connected-google-account-card.js]: Note: this is the same as the following in `AuthorizeGoogleAccountCard `, I plan to refactor this in a separate smaller PR.

https://github.com/woocommerce/google-listings-and-ads/blob/0cad7cd8defcfbd433f91ea963b12820cba41f24/js/src/components/google-account-card/authorize-google-account-card.js#L45-L58

Comment (ecgan) [reply js/src/components/google-account-card/connected-google-account-card.js]: This is addressed in https://github.com/woocommerce/google-listings-and-ads/pull/1074.

**Comment by ecgan** [comment]
@layoutd @eason9487 , just want to get a confirmation from you, for the "connect to a different Google account after already connected to an existing account" feature (see the demo video in the PR description above ☝️  ), do we need to disconnect the existing Google account first (say, by calling `DELETE /google/connect`)?

In this PR and the demo video, I did not call disconnect, and I straight away call `GET /google/connect`, and it seems to work okay. Would there be any implication on the API or WCS side? 🤔

**Comment by eason9487** [comment]
> do we need to disconnect the existing Google account first (say, by calling `DELETE /google/connect`)?

Quoted Mik's reply from https://github.com/woocommerce/google-listings-and-ads/issues/486#issuecomment-866776730:

> A disconnect request would include sending a request to the Google API to revoke the token and removes the stored token from WCS. Issuing a connect directly (even if we are already connected), will overwrite the token on WCS and leave the old Google token as is. So technically both will work, although the delete + connect would probably be better.

I think this mechanism should still be the same till now.

**Review by eason9487 — COMMENTED** [review]
If the user has only one signed in Google account, then after clicking "connect to a different Google account", they will be redirected back to GLA page directly. This makes them unable to log in to another Google account and authorize it during this process.

Kapture 2021-11-03 at 13 09 35

Seems we still need to disconnect first?

**Comment by ecgan** [comment]
@eason9487 , thanks, good findings! I'm gonna convert this PR into draft and work on disconnecting the Google account. 🙂  👍

**Comment by ecgan** [comment]
This PR is now ready for review. I have also updated the demo video in the PR description with the latest changes.

File: js/src/components/google-account-card/useSwitchGoogleAccount.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+const useSwitchGoogleAccount = () => {
+	const pageName = glaData.mcSetupComplete ? 'reconnect' : 'setup-mc';
+	const { createNotice, removeNotice } = useDispatchCoreNotices();
+	const [
+		fetchGoogleDisconnect,
+		{ loading: loadingGoogleDisconnect },
+	] = useApiFetchCallback( {
+		path: `${ API_NAMESPACE }/google/connect`,
+		method: 'DELETE',
+	} );
```
Comment (eason9487) [inline js/src/components/google-account-card/useSwitchGoogleAccount.js:45]: ❔ or 💅 

There's an action for disconnecting Google account. Are there any concerns? Or it could reuse the action here.

https://github.com/woocommerce/google-listings-and-ads/blob/7fb9c8799ab7f41789834f67c802ab396d03a305/js/src/data/actions.js#L413-L433

File: js/src/components/google-account-card/useSwitchGoogleAccount.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+ * 3. Call `GET /google/connect` API to get the Google OAuth URL.
+ * 4. Redirect the browser to the URL.
+ * 5. If there is an error in the above process, it will display an error notice.
+ *
+ * @return {Array} `[ handleSwitch, loading ]`
+ * 		- `handleSwitch` is meant to be used as button click handler.
+ * 		- `loading` is a state to indicate that the process is running.
+ */
+const useSwitchGoogleAccount = () => {
+	const pageName = glaData.mcSetupComplete ? 'reconnect' : 'setup-mc';
```
Comment (eason9487) [inline js/src/components/google-account-card/useSwitchGoogleAccount.js]: 💅 
The reconnection process will not reach here. Probably we could specify `'setup-mc'` directly.

**Review by eason9487 — APPROVED** [review]
Tested with only one signed-in Google account (https://github.com/woocommerce/google-listings-and-ads/pull/1072#pullrequestreview-796123115). Works well. LGTM.

File: js/src/components/google-account-card/useSwitchGoogleAccount.js (hunk 2 above)
Comment (ecgan) [reply js/src/components/google-account-card/useSwitchGoogleAccount.js:45]: Ahh yes, I tested that out. If we use that action, it would cause an update in the wp-data redux store, and the UI will change to the "Connect" card for a brief moment, before the browser redirects to the Google Auth page. That's why in the end I chose not to use that.

Comment (ecgan) [reply js/src/components/google-account-card/useSwitchGoogleAccount.js:45]: I added code comment for this in [`02799c3` (#1072)](https://github.com/woocommerce/google-listings-and-ads/pull/1072/commits/02799c3b90957c7e98264562500a929556b0d0f9).

File: js/src/components/google-account-card/useSwitchGoogleAccount.js (hunk 3 above)
Comment (ecgan) [reply js/src/components/google-account-card/useSwitchGoogleAccount.js]: Good call. Done in [`4910f38` (#1072)](https://github.com/woocommerce/google-listings-and-ads/pull/1072/commits/4910f3803de860184f7df2b96458007aa85feda3).

File: js/src/components/google-account-card/useSwitchGoogleAccount.js (hunk 2 above)
Comment (eason9487) [reply js/src/components/google-account-card/useSwitchGoogleAccount.js:45]: 💅 
We could add a parameter, e.g. `skipStoreInvalidation`, to the `disconnectGoogleAccount` action for skipping yielding the `TYPES.DISCONNECT_ACCOUNTS_GOOGLE`. Therefore, I think this hook would look more concise.

```js
const useSwitchGoogleAccount = () => {
	const pageName = 'setup-mc';
	const { createNotice, removeNotice } = useDispatchCoreNotices();
	const { disconnectGoogleAccount } = useAppDispatch();
	const [ fetchGoogleConnect ] = useGoogleAuthorization( pageName );
	const [ isRunning, setRunning ] = useState( false );

	const handleSwitch = async () => {
		setRunning( true );

		const { notice } = await createNotice(
			'info',
			__(
				'Connecting to a different Google account, please wait…',
				'google-listings-and-ads'
			)
		);

		try {
			await disconnectGoogleAccount( true );
			const { url } = await fetchGoogleConnect();
			window.location.href = url;
		} catch ( error ) {
			setRunning( false );
			removeNotice( notice.id );
			createNotice(
				'error',
				__(
					'Unable to connect to a different Google account. Please try again later.',
					'google-listings-and-ads'
				)
			);
		}
	};

	return [ handleSwitch, isRunning ];
};
```

Comment (ecgan) [reply js/src/components/google-account-card/useSwitchGoogleAccount.js:45]: > We could add a parameter, e.g. `skipStoreInvalidation`, to the `disconnectGoogleAccount` action for skipping yielding the `TYPES.DISCONNECT_ACCOUNTS_GOOGLE`.

I tried out the idea, and to be honest, personally I'm not really fond of that. The `skipStoreInvalidation` parameter is there to work with our wp-data technical implementation, and it does not really describe `disconnectGoogleAccount`. In other words, in layman term regardless of wp-data store, "I want to disconnect Google account, why is this skipStoreInvalidation needed?"

The other thing is that we would also need to handle it in the resolver part, like below:

```js
// add the new "! action.skipStoreInvalidation" condition below.
getGoogleAccount.shouldInvalidate = ( action ) => {
	return (
		action.type === TYPES.DISCONNECT_ACCOUNTS_GOOGLE &&
		! action.skipStoreInvalidation
	);
};
```

So to support this `skipStoreInvalidation`, we would need to modify `actions.js`, `resolvers.js` and the consuming hook here.

While it can make the hook look more concise, I think overall it also makes things harder to reason about and not as straightforward. So I'm leaning towards what we have now in here. 😄

Comment (eason9487) [reply js/src/components/google-account-card/useSwitchGoogleAccount.js:45]: Thanks for replying. I got more of your thoughts. 🙏 

> The `skipStoreInvalidation` parameter is there to work with our wp-data technical implementation, and it does not really describe `disconnectGoogleAccount`. In other words, in layman term regardless of wp-data store, "I want to disconnect Google account, why is this skipStoreInvalidation needed?"

> I think overall it also makes things harder to reason about and not as straightforward.

I think it's understandable that wp-data is the main global state manager in this project, and that adding a control parameter based on its characteristics should be a reasonable approach. Because eventually, frontend maintainers need to understand how (internal implementation of) wp-data invalidates runtime cache state and drives it to update relevant state from the source. Otherwise, we might have more and more similar workarounds.

Even if `skipStoreInvalidation` may seem a bit odd, a comment explaining its purpose is enough to show why, just as a comment is needed in this hook to explain why. And as much as possible, the state management could be kept in `. ~/data` for better codebase consistency and predictability.

> So I'm leaning towards what we have now in here.

My main point (started with ❔ ) above was to get thoughts and discussions. I think it's ok to keep this now.
